ci-operator/step-registry/cucushift/upgrade/tolatest/cucushift-upgrade-tolatest-commands.sh

```diff
@@ -0,0 +1,681 @@
+        echo "${TARGET} is signed" && return 0
+    else
+        echo "Seem like ${TARGET} is not signed" && return 1
+    fi
+}
+
+# Check if admin ack is required before upgrade
+function admin_ack() {
+    if (( SOURCE_MINOR_VERSION == TARGET_MINOR_VERSION )) || (( SOURCE_MINOR_VERSION < 8 )); then
+        echo "Admin ack is not required in either z-stream upgrade or 4.7 and earlier" && return
+    fi
+
+    local out; out="$(oc -n openshift-config-managed get configmap admin-gates -o json | jq -r ".data")"
+    if [[ ${out} != *"ack-4.${SOURCE_MINOR_VERSION}"* ]]; then
+        echo "Admin ack not required: ${out}" && return
+    fi
+
+    echo "Require admin ack:\n ${out}"
+    local wait_time_loop_var=0 ack_data
+
+    ack_data="$(echo "${out}" | jq -r "keys[]")"
+    for ack in ${ack_data};
+    do
+        # e.g.: ack-4.12-kube-1.26-api-removals-in-4.13
+        if [[ "${ack}" == *"ack-4.${SOURCE_MINOR_VERSION}"* ]]
+        then
+            echo "Admin ack patch data is: ${ack}"
+            oc -n openshift-config patch configmap admin-acks --patch '{"data":{"'"${ack}"'": "true"}}' --type=merge
+        fi
+    done
+    echo "Admin-acks patch gets started"
+
+    echo -e "sleep 5 min wait admin-acks patch to be valid...\n"
+    while (( wait_time_loop_var < 5 )); do
+        sleep 1m
+        echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+        if ! oc adm upgrade | grep "AdminAckRequired"; then
+            echo -e "Admin-acks patch PASSED\n"
+            return 0
+        else
+            echo -e "Admin-acks patch still in processing, waiting...\n"
+        fi
+        (( wait_time_loop_var += 1 ))
+    done
+    if (( wait_time_loop_var >= 5 )); then
+        echo >&2 "Timed out waiting for admin-acks completing, exiting" && return 1
+    fi
+}
+
+# check if any of cases is enabled via ENABLE_OTA_TEST
+function check_ota_case_enabled() {
+    local case_id
+    local cases_array=("$@")
+    for case_id in "${cases_array[@]}"; do
+        # shellcheck disable=SC2076
+        if [[ " ${ENABLE_OTA_TEST} " =~ " ${case_id} " ]]; then
+            echo "${case_id} is enabled via ENABLE_OTA_TEST on this job."
+            return 0
+        fi
+    done
+    return 1
+}
+
+function set_channel(){
+    x_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f1 -d. )
+    y_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f2 -d. )
+    ver="${x_ver}.${y_ver}"
+    if ! oc adm upgrade channel candidate-${ver}; then
+        echo "Fail to change channel to candidate-${ver}!"
+        return 1
+    fi
+    local retry=3
+    while (( retry > 0 ));do
+        recommends=$(oc get clusterversion version -o json|jq -r '.status.availableUpdates')
+        if [[ "${recommends}" == "null" ]]; then
+            (( retry -= 1 ))
+            sleep 60
+            echo "No recommended update available! Retry..."
+        else
+            echo "Recommencded update: ${recommends}"
+            break
+        fi
+    done
```

Comment:
looks like i left out the timeout check here, will add it together with 4.16 job later